File: src/main/java/com/qsoftware/bubbles/common/command/CommandConstructor.java
package com.qsoftware.bubbles.common.command;

import com.qsoftware.bubbles.entity.player.PlayerEntity;

import java.util.HashMap;

public final class CommandConstructor {
    protected static HashMap<String, CommandExecutor> commands = new HashMap<>();

    public static void add(String name, CommandExecutor handler) {
        commands.put(name, handler);
    }

    public static CommandExecutor get(String name) {
        if (commands.containsKey(name)) {
            return commands.get(name);
        }

        return null;
    }

    public static boolean execute(String name, PlayerEntity player, String[] args) {
        CommandExecutor handler = get(name);

        if (handler == null) {
            return false;
        }

        try {
            handler.execute(new Command(name), player, args);
        } catch (Throwable throwable) {
            player.sendMessage("Error occurred when executing the command.");
            player.sendMessage("  See the log for more information.");
            throwable.printStackTrace();
            return false;
        }
        return true;
    }
}

File: src/main/kotlin/qtech/bubbles/common/command/CommandConstructor.kt
package qtech.bubbles.common.command

import qtech.bubbles.entity.player.PlayerEntity

object CommandConstructor {
    private val commands = HashMap<String, CommandExecutor>()
    @JvmStatic
    fun add(name: String, handler: CommandExecutor) {
        commands[name] = handler
    }

    operator fun get(name: String): CommandExecutor? {
        return if (commands.containsKey(name)) {
            commands[name]
        } else null
    }

    @JvmStatic
    fun execute(name: String, player: PlayerEntity, args: Array<String>): Boolean {
        val handler = CommandConstructor[name] ?: return false
        try {
            handler.execute(player, args)
        } catch (throwable: Throwable) {
            player.sendMessage("Error occurred when executing the command.")
            player.sendMessage("  See the log for more information.")
            throwable.printStackTrace()
            return false
        }
        return true
    }
}